### PR TITLE
Add antialiasing by supersampling

### PR DESCRIPTION
**Pull request by AxelF44**
Hello Team,

I've Implemented the antialiasing throught supersampling!

Here are the result of a sphere after an antialiasing of 3:
image

And here the result whitout antialiasing:
image

### CODE REVIEW

File: tests/files_examples/subjects/rotation_test.cfg [hunk 1]
```diff
@@ -7,42 +7,48 @@ camera :
     position = { x = 0.0; y = 0.0; z = 0.0; };
     rotation = { x = 0.0; y = 0.0; z = 0.0; };
     fieldOfView = 72.0; # In degree
```
Comment (theophile-jr) [inline tests/files_examples/subjects/rotation_test.cfg:9]: degrees ?

**Review by theophile-jr — CHANGES_REQUESTED** [review]
Looks fine. Just small correction but not code ;)

File: include/Camera/Camera.hpp [hunk 2]
```diff
@@ -91,6 +91,13 @@ class Raytracer::Camera {
          */
         std::pair<double, double> getResolution(void) const;
 
+        /**
+         * @brief Get the Antialiasing object.
+         *
```
Comment (Njord201) [inline include/Camera/Camera.hpp:96]:
```suggestion
         * @brief Get the antialiasing object.
         *
```

File: include/Camera/Camera.hpp [hunk 3]
```diff
@@ -127,8 +134,15 @@ class Raytracer::Camera {
          */
         void setResolution(double width, double height);
 
-    private:
+        /**
+         * @brief Set the Antialiasing object.
```
Comment (Njord201) [inline include/Camera/Camera.hpp:138]:
```suggestion
         * @brief Set the antialiasing object.
```

File: include/Camera/CameraBuilder.hpp [hunk 4]
```diff
@@ -69,6 +69,14 @@ class Raytracer::CameraBuilder {
          */
         CameraBuilder& setResolution(double width, double height);
 
+        /**
+         * @brief Set the Antialiasing object.
```
Comment (Njord201) [inline include/Camera/CameraBuilder.hpp:73]:
```suggestion
         * @brief Set the antialiasing object.
```

File: include/Camera/CameraBuilder.hpp [hunk 5]
```diff
@@ -69,6 +69,14 @@ class Raytracer::CameraBuilder {
...
+         *
+         * @param antialiasing Value of antialiasing
+         * @return CameraBuilder& object to chain the setter
+         */
+        Raytracer::CameraBuilder&  setAntialiasing(int antialiasing);
```
Comment (Njord201) [inline include/Camera/CameraBuilder.hpp:78]:
```suggestion
        Raytracer::CameraBuilder& setAntialiasing(int antialiasing);
```

File: include/Camera/CameraBuilder.hpp [hunk 6]
```diff
@@ -69,6 +69,14 @@ class Raytracer::CameraBuilder {
...
+
         /**
          * @brief Build the Camera object.
          *
```
Comment (Njord201) [inline include/Camera/CameraBuilder.hpp:82]:
```suggestion
         *
```

File: src/Core/Renderer.cpp [hunk 7]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+    auto pixelSizeV = viewV / supersampledHeight;
 
     // Calculate the location of the upper-left pixel.
     auto viewUpper_left = _camera.getOrigin() - Math::Vector3D(0, 0, focalLength) - viewU / 2 - viewV / 2;
 
+    for (int y = supersampledHeight - 1; y >= 0; y--) {
+        for (int x = 0; x < supersampledWidth; x++) {
+            auto pixelCenter = viewUpper_left + (pixelSizeU * (x + 0.5)) + (pixelSizeV * (y + 0.5));
+            auto rayDirection = _camera.getOrigin() - pixelCenter;
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:134]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 8]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
 
+    for (int y = supersampledHeight - 1; y >= 0; y--) {
+        for (int x = 0; x < supersampledWidth; x++) {
+            auto pixelCenter = viewUpper_left + (pixelSizeU * (x + 0.5)) + (pixelSizeV * (y + 0.5));
+            auto rayDirection = _camera.getOrigin() - pixelCenter;
+ 
+            rayDirection.rotateX(this->_camera.getRotation().x());
+            rayDirection.rotateY(this->_camera.getRotation().y());
+            rayDirection.rotateZ(this->_camera.getRotation().z());
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:138]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 9]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+            auto pixelCenter = viewUpper_left + (pixelSizeU * (x + 0.5)) + (pixelSizeV * (y + 0.5));
+            auto rayDirection = _camera.getOrigin() - pixelCenter;
+ 
+            rayDirection.rotateX(this->_camera.getRotation().x());
+            rayDirection.rotateY(this->_camera.getRotation().y());
+            rayDirection.rotateZ(this->_camera.getRotation().z());
+ 
+            Raytracer::Ray r(_camera.getOrigin(), rayDirection);
+            Color hitColor = _primitives.getColorPoint(r, _lights);
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:141]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 10]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+ 
+            Raytracer::Ray r(_camera.getOrigin(), rayDirection);
+            Color hitColor = _primitives.getColorPoint(r, _lights);
+ 
+            supersampledImage[y][x] = hitColor;
+        }
+    }
+
+    std::vector<std::vector<Color>> finalImage(imageHeight, std::vector<Color>(imageWidth));
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:147]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 11]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+        }
+    }
+
+    std::vector<std::vector<Color>> finalImage(imageHeight, std::vector<Color>(imageWidth));
+ 
+    for (int y = 0; y < imageHeight; y++) {
+        for (int x = 0; x < imageWidth; x++) {
+            Color averageColor(0, 0, 0);
+            int pixelCount = 0;
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:152]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 12]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+ 
+            for (int sy = 0; sy < supersamplingFactor; sy++) {
+                for (int sx = 0; sx < supersamplingFactor; sx++) {
+                    int sourceY = y * supersamplingFactor + sy;
+                    int sourceX = x * supersamplingFactor + sx;
+                    averageColor += supersampledImage[sourceY][sourceX];
+                    pixelCount++;
+                }
+            }
+
```
Comment (Njord201) [inline src/Core/Renderer.cpp:161]:
```suggestion

```

File: src/Core/Renderer.cpp [hunk 13]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+    }
+
+    std::vector<std::vector<Color>> finalImage(imageHeight, std::vector<Color>(imageWidth));
+ 
+    for (int y = 0; y < imageHeight; y++) {
+        for (int x = 0; x < imageWidth; x++) {
+            Color averageColor(0, 0, 0);
+            int pixelCount = 0;
+ 
+            for (int sy = 0; sy < supersamplingFactor; sy++) {
```
Comment (Njord201) [inline src/Core/Renderer.cpp:153]: What does "sy" means ?

File: src/Core/Renderer.cpp [hunk 14]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+                for (int sx = 0; sx < supersamplingFactor; sx++) {
```
Comment (Njord201) [inline src/Core/Renderer.cpp:154]: What does sx means ?

**Review by Njord201 — CHANGES_REQUESTED** [review]
Some little request and two questions

**Review by mathieurobert1 — COMMENTED** [review]
Just resolves all request of @Njord201

File: tests/files_examples/subjects/rotation_test.cfg (hunk 1 above)
Comment (AxelF44) [reply tests/files_examples/subjects/rotation_test.cfg:9]: Yes in degree

File: src/Core/Renderer.cpp [hunk 15]
```diff
@@ -97,45 +97,81 @@ void Raytracer::Renderer::renderFinalScene()
...
+        }
+    }
+
+
     if (stream.is_open()) {
 
         stream << "P3\n" << imageWidth << ' ' << imageHeight << "\n255\n";
 
-        for (int y = imageHeight; y >= 0; y--) {
+        for (int y = 0; y < imageHeight; y++) {
```
Comment (mathieurobert1) [inline src/Core/Renderer.cpp:172]: Be careful, you inverse the image: 
```suggestion
        for (int y = imageHeight; y >= 0; y--) {
```

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
You inverse the image:
image
instead of :
image

**Review by Njord201 — APPROVED** [review]
Good job

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
It's still inverse, but now in the vertical way:
image
Instead of :
image

File: include/Camera/Camera.hpp [hunk 16]
```diff
@@ -127,8 +134,15 @@ class Raytracer::Camera {
          */
         void setResolution(double width, double height);
 
-    private:
+        /**
+         * @brief Set the antialiasing object.
+         *
+         * @param antialiasing Value of antialiasing
```
Comment (Njord201) [inline include/Camera/Camera.hpp:140]:
```suggestion
         * @param antialiasing value of antialiasing
```

**Review by Njord201 — CHANGES_REQUESTED** [review]

**Review by mathieurobert1 — APPROVED** [review]
Perfect this time !

**Review by mathieurobert1 — APPROVED** [review]

**Review by Njord201 — DISMISSED** [review]
Good job Mister Fradet

**Review by theophile-jr — DISMISSED** [review]

**Review by mathieurobert1 — APPROVED** [review]

**Review by Njord201 — APPROVED** [review]

**Pull request merged**